keepassxc-browser/background/keepass.js

```diff
@@ -696,6 +702,71 @@ keepass.lockDatabase = function(tab) {
     });
 };
 
+keepass.getDatabaseGroups = function(callback, tab) {
+    keepass.testAssociation((taResponse) => {
+        if (!taResponse) {
+            browserAction.showDefault(null, tab);
+            callback([]);
+            return;
+        }
+
+        if (tab && page.tabs[tab.id]) {
+            page.tabs[tab.id].errorMessage = null;
+        }
+
+        if (!keepass.isConnected) {
+            callback([]);
+            return;
+        }
+
+        let groups = [];
+        const kpAction = kpActions.GET_DATABASE_GROUPS;
+        const nonce = keepass.getNonce();
+        const incrementedNonce = keepass.incrementedNonce(nonce);
+
+        const messageData = {
+            action: kpAction
+        };
+
+        const request = {
+            action: kpAction,
+            message: keepass.encrypt(messageData, nonce),
+            nonce: nonce,
+            clientID: keepass.clientID
+        };
+
+        keepass.sendNativeMessage(request).then((response) => {
+            if (response.message && response.nonce) {
+                const res = keepass.decrypt(response.message, response.nonce);
+                if (!res) {
+                    keepass.handleError(tab, kpErrors.CANNOT_DECRYPT_MESSAGE);
+                    callback([]);
+                    return;
+                }
+
+                const message = nacl.util.encodeUTF8(res);
+                const parsed = JSON.parse(message);
+                keepass.setcurrentKeePassXCVersion(parsed.version);
+
+                if (keepass.verifyResponse(parsed, incrementedNonce)) {
+                    groups = parsed.groups;
+                    keepass.updateLastUsed(keepass.databaseHash);
+                    callback(groups);
+                } else {
+                    console.log('getDatabaseGroups rejected');
```

Comment:
Would you want to show a proper error at this point? Would the CANNOT_DECRYPT_MESSAGE be appropriate here?

Reply:
Something like "Invalid or corrupt message sent from KeePassXC"

Reply:
Something like that yes. Currently the verifyResponse() returns false from multiple reasons, so the message must be a general one. The one you proposed or something like: "Message sent from KeePassXC cannot be verified".